File: java/src/MoveThread.java
import java.util.ArrayList;
import java.util.List;

/**
 * A thread that will do move computations.
 */
public class MoveThread extends Thread {

    private BoardJob job;
    private ThreadManager threadManager;
    private int numberOfRows;
    private int numberOfTotalPegs;
    private List<Move> validMoves;

    public MoveThread(BoardJob job, ThreadManager threadManager) {
        this.job = job;
        this.threadManager = threadManager;
        this.numberOfRows = threadManager.getNumberOfRows();
        this.numberOfTotalPegs = threadManager.getNumberOfTotalPegs();
        validMoves = new ArrayList<>();
    }

    @Override
    public void run() {
        // First we populate the list of all valid moves by testing the neighbors.
        for (int i = 0 ; i < numberOfTotalPegs ; i++) {
            if (job.board[i]) {
                testNeighborMoves(i);
            }
        }

        // If we found no valid move then we are done!!!
        if (validMoves.size() == 0) {
            threadManager.foundBoard(job);
            threadManager.threadFinished(this);
            return;
        }

        // If we still have moves, then we need to enque the jobs on the thread manager's queue.
        for (Move move : validMoves) {
            boolean[] newBoard = job.board.clone();
            newBoard[move.newPosition] = true;
            newBoard[move.originalPosition] = false;
            newBoard[move.removedPiece] = false;
            BoardJob newJob =
                    new BoardJob(job.pegsLeft - 1, job.initialPeg, job.totalMoves + 1, newBoard,
                            job.moveString + "\n" + move);
            threadManager.queueJob(newJob);
        }

        threadManager.threadFinished(this);
    }

    /**
     * Finds all available moves for the peg in question and then adds them to the valid list.
     */
    private void testNeighborMoves(int pegToInspect) {
        // These are the 6 possible move directions with O being the peg in
        // question.
        //  6 1
        // 5 O 2
        //  4 3
        // We will test these in order and compile a list of the moves that are valid that the parent thread can get
        // after joining.

        int r = PegGameUtils.getRow(pegToInspect);
        int d = PegGameUtils.getDisplacement(pegToInspect);
        int land;
        int jump;

        // logger.info("Original : " + pegToInspect);

        // 1
        // ((R, D), (R - 2, D), (R - 1, D))
        land = PegGameUtils.getPegNumber(r - 2, d);
        jump = PegGameUtils.getPegNumber(r - 1, d);
        // logger.info("Test before land = " + land + " and jump = " + jump);
        testMove(new Move(pegToInspect, land, jump));

        // 2
        // ((R, D), (R, D + 2), (R, D + 1))
        land = PegGameUtils.getPegNumber(r, d + 2);
        jump = PegGameUtils.getPegNumber(r, d + 1);
        // logger.info("Test before land = " + land + " and jump = " + jump);
        testMove(new Move(pegToInspect, land, jump));

        // 3
        // ((R, D), (R + 2, D + 2), (R + 1, D + 1))
        land = PegGameUtils.getPegNumber(r + 2, d + 2);
        jump = PegGameUtils.getPegNumber(r + 1, d + 1);
        // logger.info("Test before land = " + land + " and jump = " + jump);
        testMove(new Move(pegToInspect, land, jump));

        // 4
        // ((R, D), (R - 2, D - 2), (R - 1, D - 1))
        land = PegGameUtils.getPegNumber(r + 2, d);
        jump = PegGameUtils.getPegNumber(r + 1, d);
        // logger.info("Test before land = " + land + " and jump = " + jump);
        testMove(new Move(pegToInspect, land, jump));

        // 5
        // ((R, D), (R, D - 2), (R, D - 1))
        land = PegGameUtils.getPegNumber(r, d - 2);
        jump = PegGameUtils.getPegNumber(r, d - 1);
        // logger.info("Test before land = " + land + " and jump = " + jump);
        testMove(new Move(pegToInspect, land, jump));

        // 6
        // ((R, D), (R - 2, D - 2), (R - 1, D - 1))
        land = PegGameUtils.getPegNumber(r - 2, d - 2);
        jump = PegGameUtils.getPegNumber(r - 1, d - 1);
        // logger.info("Test before land = " + land + " and jump = " + jump);
        testMove(new Move(pegToInspect, land, jump));
    }

    /**
     * Tests the move and adds it to the list of valid moves if the move is valid.
     */
    private void testMove(Move move) {
        int lR = PegGameUtils.getRow(move.newPosition);
        if (lR < 0 || lR >= numberOfRows) {
            return;
        }
        int lD = PegGameUtils.getDisplacement(move.newPosition);
        if (0 <= lR && lR <= numberOfRows - 1 && 0 <= lD && lD <= lR) {
            boolean b =
                    (job.board[move.originalPosition] && !job.board[move.newPosition] && job.board[move.removedPiece]);
            if (b) {
                validMoves.add(move);
            }
        }
    }
}

File: go/src/threading/MoveThread.go
package threading

import (
    "container/list"
    "strconv"
    "util"
)

/**
 * Holds the data that a MoveThread will have to execute with.
 */
type MoveThread struct {
    Job util.BoardJob
    ValidMoves *list.List
}

/**
 * Creates a new MoveThread for the given parameters.
 */
func NewMoveThread(job util.BoardJob) MoveThread {
    l := list.New()
    return MoveThread{job, l}
}

/**
 * Returns the string representation of a MoveThread.
 */
func MoveThreadToString(moveThread MoveThread) string {
    return "Rows : " + strconv.Itoa(util.GetNumberOfRows()) + ", Number of Total Pegs : " + strconv.Itoa(util.GetTotalPegs())
}

func RunMoveThread(moveThread MoveThread, channel chan string) {
    // First populate the move thread's list of moves.
    for i := 0 ; i < util.GetTotalPegs() ; i++ {
        if (moveThread.Job.Board[i]) {
            testNeighborMoves(moveThread, i)
        }
    }

    // If there a no valid moves then we need to notify somebody.
    if (moveThread.ValidMoves.Len() == 0) {
        channel <- util.BoardJobToString(moveThread.Job)
        return;
    }

    // If there are still moves then we will add the jobs to the job queue.
    element := moveThread.ValidMoves.Front()
    for i := 0 ; i < moveThread.ValidMoves.Len() ; i++ {
        move := element.Value.(util.Move)
        newBoard := cloneBoard(moveThread)
        newBoard[move.NewPosition] = true
        newBoard[move.OriginalPosition] = false
        newBoard[move.RemovePiece] = false
        newJob := util.NewBoardJobWithMove(moveThread.Job, newBoard, util.MoveToString(move))
        newMoveThread := NewMoveThread(newJob)
        go RunMoveThread(newMoveThread, channel)
        element = element.Next()
    }
}

/*******************
 * Private Methods *
 *******************/
func cloneBoard(moveThread MoveThread) []bool {
    newBoard := make([]bool, util.GetTotalPegs())
    for i := 0 ; i < util.GetTotalPegs() ; i++ {
        if (moveThread.Job.Board[i]) {
            newBoard[i] = true
        } else {
            newBoard[i] = false
        }
    }
    return newBoard
}


/**
 * Finds all available moves for a peg in question and adds it to the set of valid moves.
 */
func testNeighborMoves(moveThread MoveThread, pegToInspect int) {
    r := util.GetRow(pegToInspect)
    d := util.GetDisplacement(pegToInspect)

    /**
     * There are only 6 possible directions the peg can jump based on the diagram below.
     *
     *  6 1
     * 5 0 2
     *  4 3
     */

    // Direction 1
    land := util.GetPegNumber(r - 2, d)
    jump := util.GetPegNumber(r - 1, d)
    testMove(moveThread, util.Move{pegToInspect, land, jump})

    // Direction 2
    land = util.GetPegNumber(r, d + 2)
    jump = util.GetPegNumber(r, d + 1)
    testMove(moveThread, util.Move{pegToInspect, land, jump})

    // Direction 3
    land = util.GetPegNumber(r + 2, d + 2)
    jump = util.GetPegNumber(r + 1, d + 1)
    testMove(moveThread, util.Move{pegToInspect, land, jump})

    // Direction 4
    land = util.GetPegNumber(r + 2, d)
    jump = util.GetPegNumber(r + 1, d)
    testMove(moveThread, util.Move{pegToInspect, land, jump})

    // Direction 5
    land = util.GetPegNumber(r, d - 2)
    jump = util.GetPegNumber(r, d - 1)
    testMove(moveThread, util.Move{pegToInspect, land, jump})

    // Direction 6
    land = util.GetPegNumber(r - 2, d - 2)
    jump = util.GetPegNumber(r - 1, d - 1)
    testMove(moveThread, util.Move{pegToInspect, land, jump})
}

/**
  * Tests a move to see if it is valid, and if so it adds the move to the valid list.
  */
func testMove(moveThread MoveThread, move util.Move) {
    lR := util.GetRow(move.NewPosition)
    if (lR < 0 || lR >= util.GetNumberOfRows()) {
        return;
    }
    lD := util.GetDisplacement(move.NewPosition)
    if (0 <= lR && lR <= util.GetNumberOfRows() && 0 <= lD && lD <= lR) {
        b := (moveThread.Job.Board[move.OriginalPosition] && !moveThread.Job.Board[move.NewPosition] && moveThread.Job.Board[move.RemovePiece])
        if (b) {
            moveThread.ValidMoves.PushFront(move)
        }
    }
}
